pyro/poutine/runtime.py

```diff
@@ -16,6 +16,7 @@ class _DimAllocator(object):
     """
     def __init__(self):
         self._stack = []  # in reverse orientation of log_prob.shape
+        self._default_stack_idx = 0  # default stack index for allocation
```

Comment:
I like this, but this will have a side-effect of going past the outermost iarange if the internal ones are occupied. e.g. 

```python
with pyro.iarange("x", 10, dim=-2):   # dim = -2
    with pyro.iarange("y", 3):        # dim = -1
        with pyro.iarange("z", 5)     # dim = -3; threw an error earlier.
```

We were flagging this with an error in this PR, but now that I think about it, maybe this should be the right behavior.